utils/tcpbridge/connection/connection.go

```diff
@@ -0,0 +1,143 @@
+// risk of accidental conflicts.
+const StreamingPath = "/tcp-over-websocket-bridge/35218cb7-1201-4940-89e8-48d8f03fed96"
+
+// WebsocketNetConnection wraps the websocket.Conn type and adds extends it to implement the net.Conn interface
+type WebsocketNetConn struct {
+	*websocket.Conn
+	bufferedMsg []byte
+}
+
+// SetDeadline implements the net.Conn interface.
+func (c *WebsocketNetConn) SetDeadline(t time.Time) error {
+	if err := c.SetReadDeadline(t); err != nil {
+		return err
+	}
+	if err := c.SetWriteDeadline(t); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Read implements the io.Reader interface.
+func (c *WebsocketNetConn) Read(bs []byte) (count int, err error) {
+	for len(c.bufferedMsg) == 0 {
+		msgType, msg, err := c.ReadMessage()
+		if err != nil {
+			return 0, err
+		}
+		if msgType != websocket.TextMessage {
+			continue
+		}
+		rawBytes, err := hex.DecodeString(string(msg))
+		if err != nil {
+			return 0, fmt.Errorf("failure decoding a websocket message: %w", err)
+		}
+		c.bufferedMsg = rawBytes
+	}
+	for count = 0; count < len(bs) && count < len(c.bufferedMsg); count++ {
+		bs[count] = c.bufferedMsg[count]
+	}
+	c.bufferedMsg = c.bufferedMsg[count:]
+	return count, nil
+}
+
+// Write implements the io.Writer interface.
+func (c *WebsocketNetConn) Write(bs []byte) (count int, err error) {
+	if err := c.WriteMessage(websocket.TextMessage, []byte(hex.EncodeToString(bs))); err != nil {
+		return 0, fmt.Errorf("error writing a websocket message: %w", err)
+	}
+	return len(bs), nil
+}
+
+// DialWebsocket establishes a connection with the given server using websocket as the
+// underlying transport layer.
+func DialWebsocket(ctx context.Context, backendURL *url.URL, h http.Header) (net.Conn, error) {
+	conn, _, err := websocket.DefaultDialer.DialContext(ctx, backendURL.String(), h)
+	if err != nil {
+		return nil, fmt.Errorf("unable to dial websocket: %w", err)
+	}
+	return &WebsocketNetConn{Conn: conn}, nil
+}
+
+// Handler returns an HTTP handler that forwards bridged TCP connections to the given port.
+//
+// The passthroughHandler is used to forward any requests that are not part of the TCP-Over-WS bridge.
+func Handler(backendPort int, passthroughHandler http.Handler) http.Handler {
+	backendHost := fmt.Sprintf("localhost:%d", backendPort)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !websocket.IsWebSocketUpgrade(r) || r.URL.Path != StreamingPath {
+			passthroughHandler.ServeHTTP(w, r)
+			return
+		}
+		ctx, cancel := context.WithCancel(r.Context())
+		defer cancel()
+		upgrader := websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+		}
+		r = r.WithContext(ctx)
+		wsConn, err := upgrader.Upgrade(w, r, r.Header)
+		if err != nil {
+			log.Printf("Failure upgrading a websocket connection: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer wsConn.Close()
+		conn := &WebsocketNetConn{Conn: wsConn}
```

Comment:
nit: perhaps call this frontendConn to make clearer vs backendConn

Reply:
Done.